File: Chapter 10/SubwayPrinter.java
package chapter10;

import java.io.*;
import java.util.*;

public class SubwayPrinter
{
    private PrintStream out;
    
    public SubwayPrinter(OutputStream out) {
        this.out = new PrintStream(out);
    }
    
    public void printDirections(List route) {
        Connection connection = (Connection) route.get(0);
        String currentLine = connection.getLineName();
        String previousLine = currentLine;
        
        out.println("Start out at " + connection.getStation2().getName() + ".");
        out.println("Get on the " + currentLine + " heading towards " + connection.getStation2().getName() + ".");
        
        for (int i = 1; i < route.size(); i++) {
            connection = (Connection) route.get(i);
            currentLine = connection.getLineName();
            if (currentLine.equals(previousLine)) {
                out.println("  Continue past  " + connection.getStation1().getName() + "...");
            }
            else {
                out.println("When you get to " + connection.getStation1().getName() + ", get off the " + previousLine + ".");
                out.println("Switch over to the " + currentLine + ", heading towards " + connection.getStation2().getName() + ".");
                previousLine = currentLine;
            }
        }
        out.println("Get off at " + connection.getStation2().getName() + " and enjoy yourself!");
    }
}

File: Ch 10 - The OOA&D Lifecycle/Iteration 2 - Objectville Subway/Printer/SubwayPrinter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Objectville_Subway.SubwaySystem;

namespace Objectville_Subway.Printer
{
    class SubwayPrinter
    {
        public string output{get; set;}

        public void PrintDirections(List<Connection> route)
        {
            Connection currentLine = route[0];
            Connection priorLine = currentLine;

            output = "Start out at " + currentLine.Station1 + ".";
            output += "\nGet on the " + NewLineMessage(currentLine.LineName, currentLine.Station2.Name);

            for (int i = 1; i < route.Count; i++)
            {
                priorLine = currentLine;
                currentLine = route[i];

                if (currentLine.LineName == priorLine.LineName)
                {
                    output += "\n   Continue past " + currentLine.Station1.Name + " ...";
                }
                else
                {
                    // Switching lines 
                    output += "\nWhen you get to " + currentLine.Station1.Name + ", get off the " + priorLine.LineName + ".";
                    output += "\nSwitch over to the " + NewLineMessage(currentLine.LineName, currentLine.Station2.Name); ;
                }
            }

            output += "\nGet off at " + currentLine.Station2.Name + " and enjoy yourself!";
        }

        private string NewLineMessage(string nextLine, string nextStation)
        {
            return nextLine + ", heading towards " + nextStation + ".";
        }
    }
}
